### PR TITLE
Add 'secret' to the Chef DSL [beta]

### PR DESCRIPTION
**Pull request by marcparadise**
This is a proposed addition of a 'secret' helper to the Chef DSL, where
a 'secret' is private data stored within a secrets management service.  

Usage would look like the following working example:

```
  # The included ':example' secrets provider
  # requires that it be configured with a hash of secrets
  my_config = { "secret1" => "a hidden door" }

  file "/tmp/the-secret" do
    content secret(name: "secret1", service: :example, config: my_config)
  end

```
Initial constraints:

* minimal-to-no abstraction over how services handle inputs, outputs,
and errors. Each of these services has well-defined interfaces already,
and there is much more different than the same across services.  We may
revisit this as we begin implementing a range of specific secrets
fetchers.
* no caching of results.
* avoid 'provider' nomenclature. That's already well-used within Chef Infra.
  That's why a secrets 'service' has a SecretFetcher implementation and not a SecretProvider.

Security Concerns:

Because we don't provide an abstraction around the returned secret
(there is not a specific single type of returned data we can rely on),
if a secrets fetcher does not take steps to hide its own data from
incidental output (logs, debug output, exceptions) then there is a risk
of exposing this data to people and systems that have the ability to
view the output.

Signed-off-by: Marc A. Paradise <marc.paradise@gmail.com>

### CODE REVIEW

File: lib/chef/dsl/secret.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      #
+      # @example
+      #
+      # This example uses the built-in :example secret manager service, which
+      # accepts a hash of secrets.
+      #
+      #   value = secret(name: "test1", service: :example, config: { "test1" => "value1" } )
+      #   log "My secret is #{value}"
+      #
+      #   value = secret(name: "test1", service: :aws_secrets_manager, config: { "region" => "us-west-1" })
```
Comment (marcparadise) [inline lib/chef/dsl/secret.rb:41]: Oops, getting ahead of ourselves with this example

**Comment by jakauppila** [comment]
With no caching present, I'd be concerned with the potential API hits on retrieval for the services that charge on API usage, especially if secrets are consumed multiple times within the run. To reduce API usage, I'd be tempted to store results in `node.run_state` which is likely not ideal?

I assume the `SecretFetcher` will simply retrieve the value; could this optionally retrieve the secret metadata as well? For example, for AWS Secrets Manager, we specify tags that ideally we'd pull back with the value. This could be achieved with a [DescribeSecret](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_DescribeSecret.html).

Does the `SecretFetcher` handle any parsing of the secret when interacting with services? If I store a JSON blob in AWS Secrets Manager, do I simply get that blob back with the `secret` function or can I inject a json key from that blob?

**Comment by marcparadise** [comment]
These are good questions @jakauppila and we'll be exploring that space further as we expand this functionailty. 

Nothing about the current interface will limit our ability to make those kinds of implementation changes as we expand this functionality and gather feedback from usage.

File: lib/chef/secret_fetcher.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+class Chef
+  class SecretFetcher
+
+    SECRET_FETCHERS = [ :example ].freeze
+
+    # Returns a configured and validated instance
+    # of a [Chef::SecretFetcher::Base]  for the given
+    # service and configuration.
+    #
+    # @param service [Symbol] the identifier for the service that will suppor this request. Must be in
```
Comment (tas50) [inline lib/chef/secret_fetcher.rb:30]:
```suggestion
    # @param service [Symbol] the identifier for the service that will support this request. Must be in
```

**Review by tas50 — APPROVED** [review]

**Pull request merged**